File: problems/2026/04/2026-04-08-leetcode-141-环形链表/SolutionLC141.py
from typing import Optional


class ListNode:
    def __init__(self, val: int = 0, next: Optional["ListNode"] = None):
        self.val = val
        self.next = next


class Solution:
    """
    LeetCode 141：环形链表（快慢指针，O(1) 额外空间）
    """

    def has_cycle(self, head: Optional[ListNode]) -> bool:
        if head is None or head.next is None:
            return False
        slow = head
        fast = head
        while fast is not None and fast.next is not None:
            slow = slow.next
            fast = fast.next.next
            if slow is fast:
                return True
        return False


if __name__ == "__main__":
    sol = Solution()

    n1 = ListNode(1, ListNode(2, ListNode(3)))
    print("无环:", sol.has_cycle(n1))

    c = ListNode(1)
    c.next = ListNode(2)
    c.next.next = ListNode(3)
    c.next.next.next = c.next
    print("有环:", sol.has_cycle(c))

File: problems/2026/04/2026-04-08-leetcode-141-环形链表/SolutionLC141.java
/**
 * LeetCode 141：环形链表（快慢指针判环，O(1) 额外空间）
 */
class ListNode {
    int val;
    ListNode next;

    ListNode(int val) {
        this.val = val;
    }
}

public class SolutionLC141 {

    public boolean hasCycle(ListNode head) {
        // 空表或单节点无 next，不可能成环
        if (head == null || head.next == null) {
            return false;
        }
        // slow 每次 1 步，fast 每次 2 步；有环必在环内相遇
        ListNode slow = head;
        ListNode fast = head;
        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
            if (slow == fast) {
                return true;
            }
        }
        // fast 能走到 null，说明链表有尽头，无环
        return false;
    }

    public static void main(String[] args) {
        SolutionLC141 solution = new SolutionLC141();

        // 无环：1 -> 2 -> 3 -> null
        ListNode a = new ListNode(1);
        a.next = new ListNode(2);
        a.next.next = new ListNode(3);
        System.out.println("无环链表: " + solution.hasCycle(a));

        // 有环：1 -> 2 -> 3 -> 回 2
        ListNode b = new ListNode(1);
        b.next = new ListNode(2);
        b.next.next = new ListNode(3);
        b.next.next.next = b.next;
        System.out.println("有环链表: " + solution.hasCycle(b));
    }
}
